modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -763,6 +764,60 @@ static void PostOutputs(InferenceEngine::InferRequest   &request,
     }
 }
 
+class PostOutputsList {
+public:
+    PostOutputsList(size_t size,
+                    std::shared_ptr<IECallContext> ctx,
+                    std::vector<std::vector<int>>&& cached_dims);
+
+    void operator()(InferenceEngine::InferRequest &request) const;
+
+private:
+    struct Priv {
+        size_t              size;
+        std::atomic<size_t> finished{0u};
+        std::shared_ptr<IECallContext> ctx;
+        std::vector<std::vector<int>> cached_dims;
+    };
+    std::shared_ptr<Priv> m_priv;
+};
+
+PostOutputsList::PostOutputsList(size_t size,
+                                 std::shared_ptr<IECallContext> ctx,
+                                 std::vector<std::vector<int>>&& cached_dims)
+    : m_priv(new Priv()) {
+    m_priv->size = size;
+    m_priv->ctx = ctx;
+    m_priv->cached_dims = std::move(cached_dims);
+}
+
+void PostOutputsList::operator()(InferenceEngine::InferRequest &req) const {
+    auto&& ctx         = m_priv->ctx;
+    auto&& cached_dims = m_priv->cached_dims;
+    auto&& finished    = m_priv->finished;
```

Comment:
I'm not sure if making `finished` an `atomic` is enough. Not sure how this line would sync with increments, etc. Also I don't think that `if (finished == size)` is not an atomic operation. You probably need an additional lock here

Reply:
all operations with `atomics` are atomic :)
Probably to make it clear make sense to use `store`/`load` API explicitly

Reply:
You can do `if(atomic)`, but `if(atomic1 == atomic2)` won't be thread-safe in my opinion, please, correct me if I'm wrong. Anyway, discussed locally

Reply:
You're absolutely right, but in this case we compare `atomic` with const variable